### PR TITLE
subsys: usb: usb_c: Refactor Sink common code

### PR DESCRIPTION
**Pull request by sambhurst**
Move non-sink specific code into common code

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

### CODE REVIEW

File: subsys/usb/usb_c/usbc_pe_common.c [hunk 1]
```diff
@@ -720,6 +732,216 @@ void pe_drs_send_swap_exit(void *obj)
...
+
+/**
+ * @brief PE_Send_Soft_Reset Exit State
+ */
+void pe_send_soft_reset_exit(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+
+	/* Stop Sender Response Timer */
+	usbc_timer_stop(&pe->pd_t_sender_response);
```
Comment (alevkoy) [inline subsys/usb/usb_c/usbc_pe_common.c:872]: This can be another PR, but we'll be doing this a lot. Maybe we should have a superstate for SenderResponseTimer like we do downstream or at least a utility function.

File: subsys/usb/usb_c/usbc_pe_common.c [hunk 2]
```diff
@@ -789,15 +1011,20 @@ static const struct smf_state pe_states[] = {
...
+	[PE_SNK_GET_SOURCE_CAP] = SMF_CREATE_STATE(
+		pe_snk_get_source_cap_entry,
+		pe_snk_get_source_cap_run,
+		pe_snk_get_source_cap_exit,
+		NULL),
 	[PE_SEND_SOFT_RESET] = SMF_CREATE_STATE(
 		pe_send_soft_reset_entry,
 		pe_send_soft_reset_run,
-		pe_send_soft_reset_exit,
+		NULL,
```
Comment (alevkoy) [inline subsys/usb/usb_c/usbc_pe_common.c:1022]: You still need this exit function.

File: subsys/usb/usb_c/usbc_pe_snk_states.c [hunk 3]
```diff
@@ -15,6 +15,27 @@ LOG_MODULE_DECLARE(usbc_stack, CONFIG_USBC_STACK_LOG_LEVEL);
...
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	/* Initial role of sink is UFP */
+	pe_set_data_role(dev, TC_ROLE_UFP);
+
+	/* Initialize timers */
+	usbc_timer_init(&pe->pd_t_typec_sink_wait_cap, PD_T_TYPEC_SINK_WAIT_CAP_MAX_MS);
+	usbc_timer_init(&pe->pd_t_ps_transition, PD_T_SPR_PS_TRANSITION_NOM_MS);
+	usbc_timer_init(&pe->pd_t_chunking_not_supported, PD_T_CHUNKING_NOT_SUPPORTED_NOM_MS);
```
Comment (alevkoy) [inline subsys/usb/usb_c/usbc_pe_snk_states.c:32]: `ChunkingNotSupportedTimer` doesn't seem sink-specific to me.

**Review by alevkoy — CHANGES_REQUESTED** [review]

File: subsys/usb/usb_c/usbc_tc_common_internal.h [hunk 4]
```diff
@@ -13,7 +13,7 @@
 
 #include "usbc_timer.h"
 #include "usbc_stack.h"
-#include "usbc_tc.h"
+
```
[the remark this answers is not in the document]
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_tc_common_internal.h:16]: Done

File: subsys/usb/usb_c/usbc_pe_common_internal.h [hunk 5]
```diff
@@ -313,4 +323,189 @@ void policy_get_snk_cap(const struct device *dev, uint32_t **pdos, int *num_pdos
...
+ *	  should enable a different set of Source Caps. True
+ *	  is returned if new Source Caps are available, else
+ *	  false.
+ */
+bool policy_change_src_caps(const struct device *dev);
+
+/**
+ * @brief Sets the data role and updates the TCPC
+ */
+void pe_set_data_role(const struct device *dev, enum tc_data_role dr);
```
[the remark this answers is not in the document]
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common_internal.h:509]: Done

File: subsys/usb/usb_c/usbc_pe_common_internal.h [hunk 6]
```diff
@@ -313,4 +323,189 @@ void policy_get_snk_cap(const struct device *dev, uint32_t **pdos, int *num_pdos
...
+ * @param dev Pointer to the device structure for the driver instance
+ *
+ * @retval cable plug role
+ */
+enum tc_cable_plug pe_get_cable_plug(const struct device *dev);
+
+/**
+ * @brief Check if the sink request can be met by the DPM
+ */
+enum usbc_snk_req_reply_t policy_check_sink_request(
```
[the remark this answers is not in the document]
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common_internal.h:482]: Removed. They are used by the source state machine.

File: subsys/usb/usb_c/usbc_pe_common.c (hunk 1 above)
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common.c:872]: I agree and have been looking into adding a super state.

File: subsys/usb/usb_c/usbc_pe_common.c (hunk 2 above)
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common.c:1022]: Done

File: subsys/usb/usb_c/usbc_pe_snk_states.c (hunk 3 above)
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_snk_states.c:32]: Moved to common code.

File: subsys/usb/usb_c/usbc_pe_common_internal.h [hunk 7]
```diff
@@ -313,4 +323,162 @@ void policy_get_snk_cap(const struct device *dev, uint32_t **pdos, int *num_pdos
...
+ * @brief Get the current data role
+ *
+ * @param dev Pointer to the device structure for the driver instance
+ *
+ * @retval data role
+ */
+enum tc_data_role pe_get_data_role(const struct device *dev);
+
+/**
+ * @brief Sets the data role and updates the TCPC
```
Comment (alevkoy) [inline subsys/usb/usb_c/usbc_pe_common_internal.h:462]: Nit: Document parameters

**Review by alevkoy — DISMISSED** [review]

File: subsys/usb/usb_c/usbc_pe_common_internal.h (hunk 7 above)
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common_internal.h:462]: Done

**Review by alevkoy — DISMISSED** [review]

File: subsys/usb/usb_c/usbc_pe_common.c [hunk 8]
```diff
@@ -720,6 +733,216 @@ void pe_drs_send_swap_exit(void *obj)
...
+			 *      1: A SenderResponseTimer timeout occurs.
+			 *      2: OR the Protocol Layer indicates that a
+			 *         transmission error has occurred.
+			 *         (Handled in pe_report_error)
+			 */
+			else if (usbc_timer_expired(&pe->pd_t_sender_response)) {
+				pe_set_state(dev, PE_SNK_HARD_RESET);
+			}
+			break;
+		}
```
[the remark this answers is not in the document]
Comment (DianaZig) [reply subsys/usb/usb_c/usbc_pe_common.c:787]: This is an instance where the spec has a single state, but within that state it has a number of things to check off its list before we move onto the next state.  Ideally, we'll code this in a way which is both easy to read code while still easily correlating to the spec. 

It looks like we send the Accept from entry in our own code, which avoids the sub state.  I'm assuming we don't do that here to avoid sending the message when the PRL isn't running (and is that an improvement we should make in our own stack as well)?

[the remark this answers is not in the document]
Comment (alevkoy) [reply subsys/usb/usb_c/usbc_pe_common.c:787]: Since the substates are unique to this state, what if we defined a single-purpose enum for the substates and separate run functions for each substate? Would that resolve readability concerns?

[the remark this answers is not in the document]
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common.c:787]: I generally agree with you Keith but numbered states seems to be an accepted practice in Zephyr:
https://github.com/zephyrproject-rtos/zephyr/blob/7b5780e0076434a01543928245538300b44a3785/subsys/usb/device/class/dfu/usb_dfu.c#L810-L815
https://github.com/zephyrproject-rtos/zephyr/blob/7b5780e0076434a01543928245538300b44a3785/subsys/emul/emul_bmi160.c#L98-L104
https://github.com/zephyrproject-rtos/zephyr/blob/7b5780e0076434a01543928245538300b44a3785/subsys/net/lib/coap/coap.c#L318-L327

File: subsys/usb/usb_c/usbc_pe_common.c [hunk 9]
```diff
@@ -720,6 +733,216 @@ void pe_drs_send_swap_exit(void *obj)
...
+	pe->submachine = 0;
+}
+
+void pe_soft_reset_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	if (prl_is_running(dev)) {
+		switch (pe->submachine) {
```
Comment (jfischer-no) [inline subsys/usb/usb_c/usbc_pe_common.c:757]: Can you please avoid unnecessary nesting if possible?

```c

	if (!prl_is_running(dev)) {
		return;
	}

	switch (pe->submachine) {
```

File: subsys/usb/usb_c/usbc_pe_common.c [hunk 10]
```diff
@@ -720,6 +733,216 @@ void pe_drs_send_swap_exit(void *obj)
...
+	 * On entry to the PE_SNK_Chunk_Received state, the Policy Engine
+	 * Shall initialize and run the ChunkingNotSupportedTimer.
+	 */
+	usbc_timer_start(&pe->pd_t_chunking_not_supported);
+}
+
+/**
+ * @brief PE_Chunk_Received Run State
+ */
+void pe_chunk_received_run(void *obj)
```
Comment (jfischer-no) [inline subsys/usb/usb_c/usbc_pe_common.c:928]: `static void pe_chunk_received_run(void *obj)` ? as it is only used here. Please also check the other removed from [subsys/usb/usb_c/usbc_pe_snk_states_internal.h](https://github.com/zephyrproject-rtos/zephyr/pull/53582/files/1bb44993ce43f25bdd1015edca9ac5b4ec5c8ca9#diff-8cd388c1169e8169381910f8ebec9a9423d7ea6ba1d310330d56d54d3ddb6082)

**Review by jfischer-no — COMMENTED** [review]
Looks good, just few minor comments.

File: subsys/usb/usb_c/usbc_pe_common.c (hunk 10 above)
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common.c:928]: Done

File: subsys/usb/usb_c/usbc_pe_common.c (hunk 9 above)
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common.c:757]: Done

File: subsys/usb/usb_c/usbc_pe_common.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (sambhurst) [reply subsys/usb/usb_c/usbc_pe_common.c:787]: Done

**Review by desowin — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]
